docs/gni_sources.md

```diff
@@ -1,3 +1,9 @@
+### sources.gni
 
-This does not mean that you cannot use sources.gni to inject sources into existing targets, but please try to avoid it when practical.
+Use of sources.gni to include sources in upstream targets should should generally be avoided. Whenever possible try to break circular dependencies see [Recipes for Breaking Chrome Dependencies](https://www.chromium.org/developers/design-documents/cookbook/#recipes-for-breaking-chrome-dependencies) and [https://www.chromium.org/developers/design-documents/cookbook/#dependency-inversion](https://www.chromium.org/developers/design-documents/cookbook/#dependency-inversion) for examples. An interface/implementation pattern can also often be used where header files and possibly some cc files are included in the direct dependency and the code that would cause the circular dependency is included in a higher level target like `//brave/browser` to ensure that the implementation code is always linked into the final output. See [tabs:tabs_public](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/tabs/BUILD.gn;l=12) and [tabs:impl](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/tabs/BUILD.gn;l=304) and [//chrome/browser impl dependency](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/BUILD.gn;l=4354). The chromium ios code is also a good model for separating out dependencies and sometimes makes use of interface/implementation patterns.
+
+Circular dependencies can sometimes (temporarily) use `allow_circular_includes_from` to split sources up into smaller targets so they can be more easily resolved down the road. See [`//chrome/browser/ui`](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/BUILD.gn;l=3524) and [`//chrome/browser/ui`](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/BUILD.gn;l=5755) for examples.
```

Comment:
Shouldn't we use a permanent link with commit hash included because these line numbers would change from time to time? ex. https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/BUILD.gn;l=3524;drc=80bd94ca218b30eb74a107ea54b469d79b25f16d
And it seems like the first link name should be `//chrome/browser`

Reply:
good point, I'll update

Reply:
good call as some of these links are already wrong